File: interview_prep/algorithm/java17/src/test/java/hoa/can/code/AddTwoNumbersTest.java
package hoa.can.code;

import hoa.can.code.ez.AddTwoNumbers;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * Unit test for simple App.
 */
public class AddTwoNumbersTest {
    AddTwoNumbers sol = new AddTwoNumbers();

    @Test
    @DisplayName("1+2=3")
    public void test1() {
        ListNode n1 = new ListNode(1);
        ListNode n2 = new ListNode(2);

        assertEquals(
                new ListNode(3),
                sol.addTwoNumbers(n1, n2)
        );
    }

    @Test
    @DisplayName("10+20=30")
    public void test2() {
        ListNode n1 = new ListNode(1);
        n1.next = new ListNode(0);
        ListNode n2 = new ListNode(2);
        n2.next = new ListNode(0);

        ListNode r = new ListNode(3);
        r.next = new ListNode(0);
        assertEquals(
                r,
                sol.addTwoNumbers(n1, n2)
        );
    }

    @Test
    @DisplayName("99+88=187")
    public void test3() {
        ListNode n1 = new ListNode(9);
        n1.next = new ListNode(9);
        ListNode n2 = new ListNode(8);
        n2.next = new ListNode(8);

        ListNode r = new ListNode(7);
        r.next = new ListNode(8);
        r.next.next = new ListNode(1);
        assertEquals(
                r,
                sol.addTwoNumbers(n1, n2)
        );
    }


    @Test
    @DisplayName("243+564=708")
    public void test4() {
        ListNode n1 = new ListNode(2);
        n1.next = new ListNode(4);
        n1.next.next = new ListNode(3);

        ListNode n2 = new ListNode(5);
        n2.next = new ListNode(6);
        n2.next.next = new ListNode(4);

        ListNode r = new ListNode(7);
        r.next = new ListNode(0);
        r.next.next = new ListNode(8);
        assertEquals(
                r,
                sol.addTwoNumbers(n1, n2)
        );
    }
}

File: interview_prep/algorithm/Kotlin/src/test/kotlin/AddTwoNumbersTest.kt
import hoa.can.code.AddTwoNumbers
import hoa.can.code.ListNode
import org.junit.jupiter.api.Test
import kotlin.test.DefaultAsserter.assertEquals

class AddTwoNumbersTest {
    class AddTwoNumbersTest {
        var sol = AddTwoNumbers()

        @Test
        fun test1() {
            val n1 = ListNode(1)
            val n2 = ListNode(2)
            assertEquals(
                "1+2=3",
                ListNode(3),
                sol.addTwoNumbers(n1, n2)
            )
        }

        @Test
        fun test2() {
            val n1 = ListNode(1)
            n1.next = ListNode(0)
            val n2 = ListNode(2)
            n2.next = ListNode(0)
            val r = ListNode(3)
            r.next = ListNode(0)
            assertEquals(
                "10+20=30",
                r,
                sol.addTwoNumbers(n1, n2)
            )
        }

        @Test
        fun test3() {
            val n1 = ListNode(9)
            n1.next = ListNode(9)
            val n2 = ListNode(8)
            n2.next = ListNode(8)
            val r = ListNode(7)
            r.next = ListNode(8)
            r.next!!.next = ListNode(1)
            assertEquals(
                "99+88=187",
                r,
                sol.addTwoNumbers(n1, n2)
            )
        }

        @Test
        fun test4() {
            val n1 = ListNode(2)
            n1.next = ListNode(4)
            n1.next!!.next = ListNode(3)
            val n2 = ListNode(5)
            n2.next = ListNode(6)
            n2.next!!.next = ListNode(4)
            val r = ListNode(7)
            r.next = ListNode(0)
            r.next!!.next = ListNode(8)

            assertEquals(
                "243+564=708",
                r,
                sol.addTwoNumbers(n1, n2)
            )
        }
    }

}